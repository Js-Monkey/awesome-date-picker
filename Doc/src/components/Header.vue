<template>
  <div class="header">
    <div class="header-logo">
      <b-icon name="better" :font-size="60"></b-icon>
    </div>
    <div class="header-nav">
      <b-icon :font-size="14" name="left"></b-icon>
      <div class="header-nav">
        <router-link class="header-nav-item" v-for="(item,index) in routerItems" :to="`/${item.value}`"
                     :key="index">
          {{ item.name }}
        </router-link>
      </div>
      <b-icon :font-size="14" name="right"></b-icon>
    </div>
    <div class="header-github">
      <a href="https://github.com/Js-Monkey/better-datepicker" target="_blank">
        <b-icon :font-size="30" name="github1"></b-icon>
      </a>
    </div>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue'

export default defineComponent({
  name: 'Header',
  data() {
    return {
      routerItems: [
        {name: '首页', value: ''},
        {name: '安装', value: 'doc'},
      ],
    }
  }
})
</script>

<style scoped lang="scss">
@import "../assets/style/global";

$header-bg: $black;
.header {
  height: 66px;
  display: flex;
  align-items: center;
  padding: $padding-left-more;
  background: $header-bg;

  &-logo{
    cursor: pointer;
  }
  &-nav {
    margin-left: auto;
    display: flex;
    align-items: center;
    font-size: $font-size-small;

    a {
      text-decoration: none;
      color: $text-light;
    }

    &-item {
      margin: 0 9px;

      &::after {
        position: absolute;
        display: block;
        content: '';
        width: 25px;
        height: 2px;
        background-color: white;
        transition: all .2s;
        opacity: 0;
      }

      &:hover {
        &::after {
          transform: translateY(2px);
          opacity: 1;
        }
      }
    }
  }

  &-github {
    width: 100px;
    height: 100%;
    transform: skewX(-135deg) translateX(100px);
    background: $bg-2;

    svg {
      transition: .32s all;
      cursor: pointer;
      transform: skewX(-45deg) rotate(45deg) translate(10px, -4px);

      &:hover {
        transform: skewX(-45deg) rotate(53deg) translate(10px, -4px);
      }
    }
  }
}
</style>
