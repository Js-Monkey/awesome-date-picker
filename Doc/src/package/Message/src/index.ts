import {h, render} from 'vue'
import MessageComponent from './message.vue'
import {DefineComponent} from 'vue'
let timer = null
const queue = []

const Message = function (message: string): void {
  clearQueue()
  const container = document.createElement('div')
  const vm = h(MessageComponent as any, {message}) as any
  render(vm, container)
  vm.component.ctx.toggleVisibility()
  document.body.appendChild(container)
  queue.push(vm)
  timer = setTimeout(() => {
    closeMessage(vm)
  }, 1500)
}

function clearQueue() {
  clearTimeout(timer)
  queue.forEach(vm => closeMessage(vm))
}

function closeMessage(vm) {
  vm.component.ctx.toggleVisibility()

}


export default Message as unknown as DefineComponent
