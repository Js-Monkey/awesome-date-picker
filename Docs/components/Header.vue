<template>
    <div class="header" :class="{other:!atHome,'components-page':!atHome}">
        <div class="header-brand">
            <router-link :to="'/'">
              <svg
                style="width: 50px;height: 50px"
                version="1.0"
                xmlns="http://www.w3.org/2000/svg"
                xmlns:xlink="http://www.w3.org/1999/xlink"
                x="0"
                y="0"
                width="200"
                height="200"
                viewBox="0 0 100 70"
                preserveAspectRatio="xMidYMid meet"
                color-interpolation-filters="sRGB"
              >
                <g fill="#c62525" transform="translate(20,0)">
                  <svg viewBox="0 0 60 60" x="0" y="0" width="68.174252790367" height="68.174252790367">
                    <g mask="url(#5a19471d-efb8-446d-bd47-cd4c2b782712)">
                      <g>
                        <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                             viewBox="0 0 60 60" version="1.1"
                             x="0" y="0" width="60" height="60">
                          <desc>Created with Sketch.</desc>
                          <g fill-rule="evenodd">
                            <g>
                              <path d="M49.001,0 L10.999,0 C4.934,0 0,4.934 0,10.999 L0,49.001 C0,55.066 4.934,60 10.999,60 L49.001,60 C55.065,60 60,55.066 60,49.001 L60,10.999 C60,4.934 55.065,0 49.001,0"/>
                            </g>
                          </g>
                        </svg>
                      </g>
                      <g filter="url(#colors3041816095)">
                        <svg fill="#f33422" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" x="9.5" y="9.5" viewBox="12.180000305175781 7.020999908447266 75.63999938964844 85.94500732421875" enable-background="new 0 0 100 100" xml:space="preserve" width="41" height="41"><path d="M77.018,7.021c-0.586,7.217-2.695,12.479-6.352,15.788c-6.295-4.118-13.148-6.114-20.589-6.002  c-7.202-0.112-13.944,1.773-20.24,5.653c-4.452-3.42-6.561-8.571-6.337-15.438c-10.413,11.111-12.354,20.422-5.835,27.973  c-3.657,6.394-5.485,13.038-5.485,19.892c0,10.287,3.713,19.276,11.152,26.926c7.427,7.44,16.346,11.153,26.745,11.153  c10.302,0,19.152-3.713,26.591-11.153c7.439-7.649,11.152-16.639,11.152-26.926c0-7.203-1.717-13.666-5.15-19.389  C88.965,27.261,87.066,17.783,77.018,7.021z M64.496,69.123c-4.02,4.007-8.822,6.003-14.406,6.003  c-5.499,0-10.357-1.996-14.587-6.003c-4.006-4.006-6.002-8.863-6.002-14.586c0-5.709,1.996-10.567,6.002-14.573  c4.118-4.118,8.976-6.114,14.587-6.002c5.583-0.112,10.386,1.884,14.406,6.002c3.992,4.006,6.002,8.864,6.002,14.573  C70.498,60.26,68.488,65.117,64.496,69.123z"/></svg>
                      </g>
                    </g>
                    <mask id="5a19471d-efb8-446d-bd47-cd4c2b782712">
                      <g fill="white">
                        <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"
                             viewBox="0 0 60 60" version="1.1"
                             x="0" y="0" width="60" height="60">
                          <desc>Created with Sketch.</desc>
                          <g fill-rule="evenodd">
                            <g>
                              <path d="M49.001,0 L10.999,0 C4.934,0 0,4.934 0,10.999 L0,49.001 C0,55.066 4.934,60 10.999,60 L49.001,60 C55.065,60 60,55.066 60,49.001 L60,10.999 C60,4.934 55.065,0 49.001,0"/>
                            </g>
                          </g>
                        </svg>
                      </g>
                      <g fill="black">
                        <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" x="9.5" y="9.5" viewBox="12.180000305175781 7.020999908447266 75.63999938964844 85.94500732421875" enable-background="new 0 0 100 100" xml:space="preserve" width="41" height="41"><path d="M77.018,7.021c-0.586,7.217-2.695,12.479-6.352,15.788c-6.295-4.118-13.148-6.114-20.589-6.002  c-7.202-0.112-13.944,1.773-20.24,5.653c-4.452-3.42-6.561-8.571-6.337-15.438c-10.413,11.111-12.354,20.422-5.835,27.973  c-3.657,6.394-5.485,13.038-5.485,19.892c0,10.287,3.713,19.276,11.152,26.926c7.427,7.44,16.346,11.153,26.745,11.153  c10.302,0,19.152-3.713,26.591-11.153c7.439-7.649,11.152-16.639,11.152-26.926c0-7.203-1.717-13.666-5.15-19.389  C88.965,27.261,87.066,17.783,77.018,7.021z M64.496,69.123c-4.02,4.007-8.822,6.003-14.406,6.003  c-5.499,0-10.357-1.996-14.587-6.003c-4.006-4.006-6.002-8.863-6.002-14.586c0-5.709,1.996-10.567,6.002-14.573  c4.118-4.118,8.976-6.114,14.587-6.002c5.583-0.112,10.386,1.884,14.406,6.002c3.992,4.006,6.002,8.864,6.002,14.573  C70.498,60.26,68.488,65.117,64.496,69.123z"/></svg>
                      </g>
                    </mask>
                  </svg>
                  <defs>
                    <filter id="colors3041816095">
                      <feColorMatrix type="matrix"
                                     values="0 0 0 0 0.7734375  0 0 0 0 0.14453125  0 0 0 0 0.14453125  0 0 0 1 0"
                      />
                    </filter>
                    <filter id="colorsf3248718108">
                      <feColorMatrix type="matrix"
                                     values="0 0 0 0 0.99609375  0 0 0 0 0.99609375  0 0 0 0 0.99609375  0 0 0 1 0"
                      />
                    </filter>
                    <filter id="colorsb3584359014">
                      <feColorMatrix type="matrix" values="0 0 0 0 0  0 0 0 0 0  0 0 0 0 0  0 0 0 1 0"
                      />
                    </filter>
                  </defs>
                </g>
              </svg>
            </router-link>
        </div>
        <div class="header-nav">
            <f-icon :color="fColor" name="left"></f-icon>
            <router-link class="header-nav-item" :to="`/${item.value}`" v-for="(item,index) in routerItems"
                         :key="index">
                {{item.name}}
            </router-link>
            <f-icon :color="fColor" name="right"></f-icon>
        </div>
        <div :class="{'header-angle':atHome}">
            <a href="https://github.com/jumodada/date-picker" target="_blank">
                <f-icon :color="fColor"
                        :font-size="35"
                        :class="atHome?'Header-angle-icon-home':'Header-angle-icon-other'" name="github1">
                </f-icon>
            </a>
        </div>
    </div>
</template>

<script>
    export default {
        name: "Header",
        computed:{
            fColor(){
                return this.atHome?'white':'#ffb311'
            }
        },
        data() {
            return {
                routerItems: [
                    {name: '首页', value: ''},
                    {name: '安装', value: 'components'},
                ],
                atHome: true
            }
        },
        mounted() {
            this.isHome()
        },
        methods: {
            isHome() {
                this.atHome = this.$route.name === 'home'
            }
        },
        watch: {
            '$route'() {
                this.isHome()
            }
        }
    }
</script>

<style scoped lang="scss">
    @import "../assets/styles/color/color-select";
    @import "../assets/styles/shadow-select";

    .header {
        position: relative;
        background-color: $header-dark-2;
        z-index: 9999;
        width: 100%;
        padding: 19px;
        display: flex;
        transition: 0.36s color ease;
        align-items: center;
        -webkit-user-select: none;
        -moz-user-select: none;
        -ms-user-select: none;
        user-select: none;
        &-brand {
            cursor: pointer;
            float: left;
        }

        &-nav {
            width: 100%;
            display: inline-flex;
            align-items: center;
            justify-content: flex-end;
            text-align: end;
            margin-right: 120px;

            &-item {
                margin-left: 15px;
                margin-right: 15px;
                color: white;
                cursor: pointer;
                text-decoration: none;
                transition: color .38s ease;

                &::after {
                    position: absolute;
                    display: block;
                    content: '';
                    width: 33px;
                    height: 2px;
                    background-color: white;
                    transition: all .2s;
                    opacity: 0;
                }

                &:hover {
                    &::after {
                        transform: translateY(2px);
                        opacity: 1;
                    }
                }
            }
        }

        &-angle {
            position: absolute;
            right: 0;
            top: 0;
            width: 120px;
            height: 120px;
            border-width: 60px;
            border-style: solid;
            border-color: #1d2022 #1d2022 #313639 #313639;
            background-color: white;

            &-icon-home {
                cursor: pointer;
                position: absolute;
                top: -40px;
                transform: rotate(45deg);
                transition: .25s all ease;

                &:hover {
                    transform: rotate(56deg) scale(1.05);
                }
            }

            &-icon-other {
                cursor: pointer;
                transition: .25s all ease;

                &:hover {
                    transform: rotate(6deg) scale(1.05);
                }
            }
        }
    }

    .other {
        color: $text1-orange;
        padding-top: 16px;
        padding-bottom: 16px;
        box-shadow: $shadow-down-white;

        .header-nav {
            margin-right: 85px;
            transform: translateY(1px);
            &-item {
                color: $text1-orange;
                &::after {
                    position: absolute;
                    display: block;
                    content: '';
                    width: 33px;
                    height: 2px;
                    background-color: white;
                    transition: all .2s;
                    opacity: 0;
                }

                &:hover {
                    &::after {
                        transform: translateY(2px);
                        opacity: 1;
                    }
                }
            }
        }
    }
    .components-page{
        background-color: $header-dark-1;
    }
</style>
