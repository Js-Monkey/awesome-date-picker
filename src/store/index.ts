import {dependence, depWatcher, StateExtends} from '../types/store'
import initState from './state'
import {_Event} from '../types/event'

const Store = (function () {
  let uid = 0
  let states = [] as any[]

  function get(key: keyof StateExtends) {
    return states[uid][key]
  }

  function set(key: keyof StateExtends, val: unknown) {
    states[uid][key] = val
  }

  function addDep(key: keyof depWatcher, dep: dependence) {
    states[uid].deps[key].push(dep)
  }

  // function removeNull(): void {
  //   states = states.filter(state => Object.keys(state).length > 0) //state滤除空对象
  // }

  function createState(): void {
    states.push(initState())
    uid = states.length - 1
  }

  function changeUID(e: _Event) {
    const el = e.target || e
    states = states.filter(s => Object.keys(s).length > 0)
    uid = states.findIndex(s => s.reference === el)
  }

  function closeAllButHasId(): void {
    states.forEach((s, idx) => {
      if (idx !== uid && s.popover) {
        s.visible = false
      }
    })
  }

  function openPopover(e: _Event): void {
    changeUID(e)
    closeAllButHasId()
    if (states[uid].visible) return
    states[uid].visible = true
  }

  return {get, set, changeUID, openPopover, createState, addDep}
})()

export const {get, set, changeUID, openPopover, createState, addDep} = Store
