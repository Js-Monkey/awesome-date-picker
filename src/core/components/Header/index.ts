import {HeaderType} from '../../../types/components'
import {createElement} from '../../../utils/element'
import {header, defaultCursor} from '../../../utils/classes'
import {nextMonth, nextYear, preMonth, preYear, toggleVisibility, toMonthPage, toYearPage} from '../utils'
import {pageName, State} from '../../../types/store'
import {getMinInTen} from '../../../utils/date'
import {CreateElementOptions} from '../../../types/utils'
import {Bind} from "../../../utils/bind"
import {Sub} from "../../../types/observer";

let type: HeaderType = 'start'

const togglePage = {
  key: ['page'],
  cb: toggleVisibility
}

function yearRange(state: State) {
  const range = (year: number) => {
    const min = getMinInTen(year)
    const max = min + 9
    return min + '年' + ' - ' + max + '年'
  }
  return createElement(
    {
      name: 'span',
      text: {
        key: {
          name: 'start',
          childKey: ['year', '_day']
        },
        cb: (year: number) => range(year)
      },
      visible: {
        key: ['page'],
        cb: (page: pageName) => page === 'year'
      },
      event: toYearPage
    },
    state
  )
}

function year(state: State) {
  return createElement({
    name: 'span',
    text: {
      key: {
        name: type,
        childKey: ['year', '_day']
      },
      cb: year => year + '年'
    },
    event: toYearPage,
    visible: {
      key: ['page'],
      cb: (page: pageName) => page !== 'year'
    }
  }, state)
}

function month(state: State) {
  return createElement({
    name: 'span',
    text: {
      key: {
        name: type,
        childKey: ['month']
      },
      cb: month => month + '月'
    },
    event: toMonthPage,
    visible: togglePage
  }, state)
}

function date(state: State) {
  const showMonth: Sub<string> = {
    key: {
      name: type,
      childKey: ['year']
    },
    cb: (year) => year + '年 '
  }

  const showDate: Sub<string> = {
    key: {
      name: type,
      childKey: ['month', 'year']
    },
    cb: (month, year) => year + '年 ' + month + '月'
  }
  const isMonth = state.options.type.indexOf('month') > -1
  return createElement({
    name: 'span',
    text: isMonth ? showMonth : showDate,
    class: [defaultCursor]
  }, state)
}

function preYearIcon(state: State) {
  return createElement(
    {
      name: 'svg',
      text: 'year',
      style: {
        position: 'absolute',
        left: '30px',
        width: '14px',
        height: '14px',
      },
      event: preYear
    },
    state
  )
}

function preMonthIcon(state: State) {
  return createElement(
    {
      name: 'svg',
      text: 'month',
      style: {
        position: 'absolute',
        left: '50px',
        width: '14px',
        height: '14px',
      },
      event: preMonth,
      visible: togglePage
    },
    state
  )
}

function nextYearIcon(state: State) {
  const opt: CreateElementOptions = {
    name: 'svg',
    text: 'year',
    style: {
      position: 'absolute',
      right: '30px',
      transform: 'rotate(180deg)',
      width: '14px',
      height: '14px',
    },
    event: Bind(nextYear, type)
  }
  return createElement(opt, state)
}

function nextMonthIcon(state: State) {
  return createElement(
    {
      name: 'svg',
      text: 'month',
      style: {
        position: 'absolute',
        right: '50px',
        transform: 'rotate(180deg)',
        width: '14px',
        height: '14px',
      },
      event: Bind(nextMonth, type),
      visible: togglePage
    },
    state
  )
}

const headerChildren = {
  start: [preYearIcon, preMonthIcon, date],
  main: [preYearIcon, preMonthIcon, yearRange, year, month, nextYearIcon, nextMonthIcon],
  end: [date, nextYearIcon, nextMonthIcon]
}

export function Header(state: State, t?: HeaderType): Node {
  const opt = {
    class: [header],
    children: headerChildren[t || 'main'],
    style: {
      width: '298px',
      'text-align': 'center'
    }
  }
  type = t || 'start'
  return createElement(opt, state)
}

export function HeaderLeft(state: State): Node {
  return Header(state, 'start')
}

export function HeaderRight(state: State): Node {
  return Header(state, 'end')
}
