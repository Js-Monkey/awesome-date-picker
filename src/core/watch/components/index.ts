import {addWatch} from '../../../observer/watcher'
import {State} from '../../../types/store'
import {on} from '../../../utils/event'
import clickOutside from '../../../utils/clickoutside'
import {listenToScrollParents} from '../../../utils/listenToParents'
import {set} from '../../../store'
import {createPopover} from '../../dom/create-popover'
import Options from '../../../types/options'
import {isInBody} from '../../../utils/isInBody'
import {appendChild} from '../../../utils/element'
import {setPopoverStyle, updatePopover} from '../../dom/popover-style'

export function watchComponents(): void {
  addWatch({
    key: ['reference'],
    cb(ref: HTMLElement, state: State): void {
      on(ref, () => (state.visible = true))
      on(document.body, clickOutside.bind(null, state as any))
      listenToScrollParents(ref, state)
      set('popover', createPopover(state))
    }
  })
  addWatch({
    key: ['popover', 'options'],
    cb(pop: HTMLElement, options: Options): void {
      if (!isInBody(pop)) {
        appendChild(pop as Element)
        const {zIndex} = options
        setPopoverStyle(pop as HTMLElement, zIndex as number)
      }
    }
  })
  addWatch({
    key: ['popover', 'visible'],
    cb(pop: HTMLElement, show: boolean, state: State): void {
      updatePopover(pop, show, state)
    }
  })
}
