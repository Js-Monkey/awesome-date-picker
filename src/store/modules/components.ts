import {DayComponents, stateComponent} from '../../types/store'
import {observe} from '../../observer'

const dayComponents = Array.from({length: 42}).map((_: unknown) => {
  return observe<DayComponents>({
    text: '',
    status: 'normal'
  })
})

export default function (): stateComponent {
  return {
    reference: null,
    popover: null,
    startDayComponents: dayComponents
  }
}
