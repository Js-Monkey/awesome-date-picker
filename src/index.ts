import BD from './core'
import Options from './types/options'
import {BetterPickerInstance} from "./types/core"
import {changeOpt} from "./core/util/default-options"
import {pickerLocale} from "./store/modules/options"
import {destroyed} from "./store"

export function createDatePicker(el: HTMLInputElement, options?: Options): BetterPickerInstance | undefined {
  const picker = BD()
  return picker(el, options)
}

export const defaultOptions =  changeOpt

export const locale =  pickerLocale

export const destroy = destroyed
