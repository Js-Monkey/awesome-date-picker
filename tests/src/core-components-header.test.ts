import {createState} from "../../src/store"
import defaultOptions from "../../src/core/util/default-options"
import {HeaderLeft, HeaderRight} from "../../src/core/components/Header"

const state = createState(defaultOptions())

describe('HeaderLeft',()=>{
  it('should return HTMLElement',()=>{
    expect(HeaderLeft(state)).toMatchSnapshot()
  })
})

describe('HeaderRight',()=>{
  it('should return HTMLElement',()=>{
    expect(HeaderRight(state)).toMatchSnapshot()
  })
})
