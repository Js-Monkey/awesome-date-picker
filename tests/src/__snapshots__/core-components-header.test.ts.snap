// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderLeft should return HTMLElement 1`] = `
Object {
  "children": Array [
    [Function],
    [Function],
    [Function],
  ],
  "class": Array [
    "header",
  ],
  "style": Object {
    "text-align": "center",
    "width": "298px",
  },
}
`;

exports[`HeaderRight should return HTMLElement 1`] = `
Object {
  "children": Array [
    [Function],
    [Function],
    [Function],
  ],
  "class": Array [
    "header",
  ],
  "style": Object {
    "text-align": "center",
    "width": "298px",
  },
}
`;
